test/test_broker.hpp

```diff
@@ -746,7 +907,38 @@ class test_broker {
                         std::make_tuple(item.topic, d.contents, *(d.props))
                         );
                 }
-                subs_.emplace(item.topic, spep, item.qos_value, item.rap_value);
+                subs_.emplace(item.share, item.topic, spep, item.qos_value, item.rap_value);
+                // Since we are re-activating this session, any shared-subscriptions
+                // that are in the saved_shared_subs_ list that this client_id belongs to
+                // should be sent to the list of subscribed clients (Which will be only this client)
+                // and should then be removed from the saved_shared_subs_ list, since now it's active.
+                // Note: Even if there are other deactivated sessions in the saved_subs_ list,
+                // because even one session subscribed to the shared subscriotion is active
+                // it should not be stored in the saved subs list.
+                if(!item.share.empty()) {
+                    auto & sharedSubsIdx = saved_shared_subs_.get<tag_combined>();
+                    auto const& it = sharedSubsIdx.find(std::make_tuple(item.share, item.topic));
+                    if(it != sharedSubsIdx.end()) {
+                        // Send the saved messages out on the wire.
+                        for (auto & d : it->messages) {
+                            // But *only* for this connection
+                            // Not every connection in the broker.
+                            ep.publish(
+                                as::buffer(item.topic),
+                                as::buffer(d.contents),
+                                // TODO: why is this 'retain'?
+                                std::min(item.qos_value, d.qos_value) | MQTT_NS::retain::yes,
+                                *(d.props),
+                                std::make_tuple(item.topic, d.contents, *(d.props))
+                                );
+                        }
+                        sharedSubsIdx.erase(it);
+                    }
+                }
+
+                // This initializes the round-robin iterator storage for this shared subscription, if and only if, it is not already.
+                // Notably, we do not check the return of the emplace call, since if the item already existed, we don't want a new one made.
+                shared_subs_.emplace(item.share, item.topic, subs_.get<tag_topic>().find(std::make_tuple(item.share, item.topic)));
```

Comment:
When I set a breakpoint here, then stopped here on failure test case. The test case doesn't use shared subscriptions.

```
(gdb) p item
$10 = (const test_broker::session_subscription &) @0x7ffff0003870: {
  client_id = {
    <std::basic_string_view<char, std::char_traits<char> >> = "cid1",
    members of mqtt::buffer:
    lifetime_ = count 2, weak count 1 = {
      value = 0 '\000'
    }
  },
  share = {
    <std::basic_string_view<char, std::char_traits<char> >> = "",
    members of mqtt::buffer:
    lifetime_ = uninitialized
  },
  topic = {
    <std::basic_string_view<char, std::char_traits<char> >> = "topic1",
    members of mqtt::buffer:
    lifetime_ = count 2, weak count 1 = {
      value = 0 '\000'
    }
  },
  messages = std::vector of length 0, capacity 0,
  qos_value = mqtt::qos::at_least_once,
  rap_value = mqtt::rap::dont
}
```

I will continue debugging.

Reply:
I think maybe problem is that this code block needs to be inside the if(!item.share.empty()) check.

shared_subs_ should have no items with empty share.

Quote (https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901250): A Shared Subscription's Topic Filter MUST start with $share/ and MUST contain a ShareName that is at least one character long

Because this is being initialized, it causes this code, in the do_publish() function (https://github.com/redboltz/mqtt_cpp/pull/672/files/3663a61452f2a3aaad022a23a41f6267a1ef4759#diff-45ca9626e6c7c611be2ceb2ff63031a5R1232) to send duplicate messages.

Reply:
I've pushed a change that may fix the issue.